src/index.ts

```diff
@@ -0,0 +1,449 @@
+      this.publicKey = ethUtil.privateToPublic(this.privateKey as Buffer)
+    }
+    return this.publicKey
+  }
+
+  private get privKey(): Buffer {
+    if (!keyExists(this.privateKey)) {
+      throw new Error('This is a public key only wallet')
+    }
+    return this.privateKey
+  }
+
+  constructor(
+    private readonly privateKey?: Buffer | undefined,
+    private publicKey: Buffer | undefined = undefined,
+  ) {
+    if (privateKey && publicKey) {
+      throw new Error('Cannot supply both a private and a public key to the constructor')
+    }
+
+    if (!privateKey && !publicKey) {
+    }
+
+    if (privateKey && !ethUtil.isValidPrivate(privateKey)) {
+      throw new Error('Private key does not satisfy the curve requirements (ie. it is invalid)')
+    }
+
+    if (publicKey && !ethUtil.isValidPublic(publicKey)) {
+      throw new Error('Invalid public key')
+    }
+  }
+
+  // public instance methods
+
+  public getPrivateKey(): Buffer {
+    return this.privKey
+  }
+
+  public getPrivateKeyString(): string {
+    return ethUtil.bufferToHex(this.privKey)
+  }
+
+  public getPublicKey(): Buffer {
+    return this.pubKey
+  }
+
+  public getPublicKeyString(): string {
+    return ethUtil.bufferToHex(this.getPublicKey())
+  }
+
+  public getAddress(): Buffer {
+    return ethUtil.publicToAddress(this.pubKey)
+  }
+
+  public getAddressString(): string {
+    return ethUtil.bufferToHex(this.getAddress())
+  }
+
+  public getChecksumAddressString(): string {
+    return ethUtil.toChecksumAddress(this.getAddressString())
+  }
+
+  public toV3(password: string, opts?: Partial<V3Params>) {
+    if (!keyExists(this.privateKey)) {
+      throw new Error('This is a public key only wallet')
+    }
+
+    const params = mergeToV3ParamsWithDefaults(opts)
+    const kdfParams = mergeKDFParamsWithDefaults(params)
+
+    let derivedKey: Buffer, finalKDFParams: Partial<KDFParams>
+    if (params.kdf === 'pbkdf2') {
+      derivedKey = crypto.pbkdf2Sync(
+        Buffer.from(password),
+        params.salt,
+        kdfParams.c,
+        kdfParams.dklen,
+        'sha256',
+      )
+      finalKDFParams = stripUnusedKDFParamsForPBKDF2(kdfParams)
+    } else if (params.kdf === 'scrypt') {
+      // FIXME: support progress reporting callback
+      derivedKey = scryptsy(
+        Buffer.from(password),
+        params.salt,
+        kdfParams.n,
+        kdfParams.r,
+        kdfParams.p,
+        kdfParams.dklen,
+      )
+      finalKDFParams = stripUnusedKDFParamsForScrypt(kdfParams)
+    } else {
+      throw new Error('Unsupported kdf')
+    }
+
+    const cipher: crypto.Cipher = crypto.createCipheriv(
+      params.cipher,
+      derivedKey.slice(0, 16),
+      params.iv,
+    )
+    if (!cipher) {
+      throw new Error('Unsupported cipher')
+    }
+
+    const ciphertext = runCipherBuffer(cipher, this.privKey)
+    const mac = ethUtil.keccak256(
+      Buffer.concat([derivedKey.slice(16, 32), Buffer.from(ciphertext)]),
+    )
+
+    return {
```

Comment:
Return value of `toV3` could use the same type annotation as `fromV3`'s input type

Reply:
Agreed